feat: реализовать HTTP-сервер и обработчики для задач (согласно ТЗ_9)

cepheil opened the pull request:
server:

Реализовать классы-обработчики HTTP-запросов:
-BaseHttpHandler
-TasksHttpHandler
-SubtasksHttpHandler
-EpicsHttpHandler
-HistoryHttpHandler
-PrioritizedHttpHandler
Добавить основной серверный класс: HttpTaskServer

util:
Добавить адаптеры для сериализации/десериализации: -LocalDateTimeAdapter
-DurationAdapter

core:
Выполнен рефакторинг CustomFormatter
Обновлены модели Task и Subtask для поддержки новых требований API

test/server:
Написать и выполнить тесты для обработчиков:
-TasksHttpHandlerTest
-SubtasksHttpHandlerTest
-EpicsHttpHandlerTest
-HistoryHttpHandlerTest
-PrioritizedHttpHandlerTest

cepheil committed:
feat: реализовать HTTP-сервер и обработчики для задач (согласно ТЗ_9)
server:

Реализовать классы-обработчики HTTP-запросов:
-BaseHttpHandler
-TasksHttpHandler
-SubtasksHttpHandler
-EpicsHttpHandler
-HistoryHttpHandler
-PrioritizedHttpHandler
Добавить основной серверный класс: HttpTaskServer

util:
Добавить адаптеры для сериализации/десериализации:
-LocalDateTimeAdapter
-DurationAdapter

core:
Выполнен рефакторинг CustomFormatter
Обновлены модели Task и Subtask для поддержки новых требований API

test/server:
Написать и выполнить тесты для обработчиков:
-TasksHttpHandlerTest
-SubtasksHttpHandlerTest
-EpicsHttpHandlerTest
-HistoryHttpHandlerTest
-PrioritizedHttpHandlerTest

Signed-off-by: Sergei Levashov <cephei.l@yandex.ru>

avfyodorov commented on src/server/HttpTaskServer.java:
```diff
@@ -0,0 +1,73 @@
+    private final TaskManager taskManager;
+    private static final int PORT = 8080;
+    private HttpServer httpServer;
+    private static final Gson gson = new GsonBuilder()
```
Можно было бы вынести  создание Gson  в базовый  обработчик, в класс BaseHttpHandler .   
Тогда не пришлось бы передавать его в качестве входного аппарата. 
На Ваше усмотрение.

avfyodorov commented on src/server/SubtasksHttpHandler.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
```
Можно было бы перенести этот метод базовый класс, а также написать там методы заглушки вида 
```java
public  class BaseHandler implements HttpHandler {
......................

public void handle(HttpExchange exchange) throws IOException { 
       String method = exchange.getRequestMethod();
           switch (method) {
               case "GET":
                   processGet(exchange);
                   break;
               case "POST":
                   processPost(exchange);
                   break;
               case "DELETE":
                   processDelete(exchange);
                   break;
               default:
                   writeToUser(exchange, "Данный метод не предусмотрен");
           }
           
    protected void processGet(....) {
        sendMethodNotAllowed(...);
   }
   protected void processPost(....) {
        sendMethodNotAllowed(...);
   }
.......................
И здесь же общие методы для классов-наследников. 

```
Тогда в обработчиках-наследниках пришлось бы переписывать только эти конкретные методы.

Для задач приблизительно так:
```java
public class TasksHandler extends BaseHandler {
......................
    public TasksHandler(........................
    
    @Override
    protected void processGet(HttpExchange exchange, String path) throws IOException {
  .........................
    @Override
    protected void processPost(HttpExchange exchange) throws IOException {
 ...............
    @Override
    protected void processDelete(HttpExchange exchange, String path) throws IOException {
.......................
```
.-----------------------------

А для отсортированного по времени списка достаточно будет переопределить всего один метод: 

```java
public class PrioritizedHandler extends BaseHandler {

    @Override
    protected void processGet(HttpExchange exchange, String path) throws IOException {
..................
```
.---------------------

Аналогично определяются классы-наследники и для эпиков, подзадач, истории. 

На Ваше усмотрение.

avfyodorov reviewed (commented):
Добрый вечер, Сергей! 

Отличная работа, всего пара уточнений, да и  они на Ваше усмотрение. 
Если не будете  менять код, просто ещё раз перешлите работу.

cepheil committed:
refactor: вынести общую логику в BaseHttpHandler, упростить обработчики

- Перенести реализацию метода handle() в BaseHttpHandler для устранения дублирования
- Перенести gson в BaseHttpHandler и сделать общим для всех обработчиков
- Обновить классы: EpicsHttpHandler, HistoryHttpHandler, HttpTaskServer, PrioritizedHttpHandler, SubtasksHttpHandler, TasksHttpHandler
- Удалить дублирующий код и лишние вызовы exchange.close()

Signed-off-by: Sergei Levashov <cephei.l@yandex.ru>

avfyodorov reviewed (commented):
Добрый день, Сергей!

Замечаний нет.
Работа принята.
